hack/generator/pkg/astbuilder/builder.go

```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package astbuilder
+
+import (
+	"fmt"
+	"go/ast"
+	"go/token"
+)
+
+// CheckErrorAndReturn checks if the err is non-nil, and if it is returns.
+func CheckErrorAndReturn(otherReturns ...*ast.Ident) ast.Stmt {
+
+	var returnValues []ast.Expr
+	for _, ret := range otherReturns {
+		returnValues = append(returnValues, ret)
+	}
+	returnValues = append(returnValues, ast.NewIdent("err"))
+
+	return &ast.IfStmt{
+		Cond: &ast.BinaryExpr{
+			X:  ast.NewIdent("err"),
+			Op: token.NEQ,
+			Y:  ast.NewIdent("nil"),
+		},
+		Body: &ast.BlockStmt{
+			List: []ast.Stmt{
+				&ast.ReturnStmt{
+					Results: returnValues,
+				},
+			},
+		},
+	}
+}
+
+// NewStruct creates a new assignment statement where a struct is constructed and stored in a variable of the given name
+func NewStruct(varName *ast.Ident, structName *ast.Ident) ast.Stmt {
+	return &ast.AssignStmt{
+		Lhs: []ast.Expr{
+			varName,
+		},
+		Tok: token.DEFINE,
+		Rhs: []ast.Expr{
+			&ast.CompositeLit{
+				Type: structName,
+			},
+		},
+	}
+}
+
+// Check if a variable is nil and if it is, return
+func NilGuardAndReturn(toCheck *ast.Ident, returns ...*ast.Ident) ast.Stmt {
+
+	var returnValues []ast.Expr
+	for _, ret := range returns {
+		returnValues = append(returnValues, ret)
+	}
+
+	return &ast.IfStmt{
+		Cond: &ast.BinaryExpr{
+			X:  toCheck,
+			Op: token.EQL,
+			Y:  ast.NewIdent("nil"),
+		},
+		Body: &ast.BlockStmt{
+			List: []ast.Stmt{
+				&ast.ReturnStmt{
+					Results: returnValues,
+				},
+			},
+		},
+	}
+}
+
+// DefineFunc defines a function (header, body, etc)
+func DefineFunc(
+	receiverIdent *ast.Ident,
+	receiverType ast.Expr,
+	name *ast.Ident,
+	comment string,
+	params []*ast.Field,
+	returns []*ast.Field,
+	body []ast.Stmt) *ast.FuncDecl {
+
+	// TODO: I'm not actually sure this is better than just declaring it all inline...?
+	// TODO: Maybe if this took a struct param instead of so many individual items?
```

Comment:
I think a parameter struct would definitely be good here, doubly so if some of the args are optional.

Reply:
With the way this is written right now, all of the args are required. I've updated it to where the comment is optional though and moved to a struct paramter.